Mk 243262 migration for new location fixture

mkangia opened the pull request:
[Spec](https://docs.google.com/document/d/1KjGPZL4oHyFUeAajnyNYII5Un276_ofc8WiTbWkoaYM/edit#)

🏛 

command to run 
`./manage.py migrate_new_location_fixture`
for dry run use `./manage.py migrate_new_location_fixture --check`

mkangia committed:
Add migration for new location fixture, remove implementation for Feature Flag [skip ci]

mkangia committed:
Reveal Fixture configuration when HIERARCHICAL_LOCATION_FIXTURE FF enabled

mkangia committed:
Update migration to fetch domains with applications using old location fixture

mkangia committed:
Update conditions to use flat/hierarchical location fixture

mkangia committed:
Add note to self

mkangia committed:
fix search string to match in applications docs

mkangia committed:
fix condition to use hierarchical fixture by including configuration

mkangia commented:
@esoergel would add more checks for finding out applications using hierarchical fixture. But everything else should be in place. Updated Spec as well.

proteusvacuum commented on corehq/apps/locations/management/commands/migrate_new_location_fixture.py:
```diff
@@ -0,0 +1,40 @@
+        domains_with_hierarchical_fixture = find_applications_with_hierarchical_fixture(
+            domains_having_locations
+        )
+        toggle = Toggle.get(HIERARCHICAL_LOCATION_FIXTURE.slug)
```
I think here you want to do `HIERARHICAL_LOCATION_FIXTURE.set(domain, enabled=True, namespace=NAMESPACE_DOMAIN)`, because that also updates the toggle cache

mkangia replied:
need to correct that.. i intended to use `toggle.set` only.

czue replied:
you can also just append the (namespaced) domain to the list of enabled items and save it at the end if we don't want to save this document once-per domain. you can just look at what `set_toggle` does behind the scenes

mkangia replied:
Ideally would do that. I am not completely sure but seems like the cache expiry would get tricky there since that needs to be done for each domain.

proteusvacuum commented on corehq/apps/locations/management/commands/migrate_new_location_fixture.py:
```diff
@@ -0,0 +1,40 @@
+    for domain in domains:
+        domain_obj = Domain.get_by_name(domain)
+        for application in domain_obj.applications():
+            raw_doc = json.dumps(application.get_db().get(application.id))
```
does this also find references in form xml?

mkangia replied:
Not yet Farid. Waiting for some feedback for more areas to search. Included application doc only for now.

czue replied:
I think the rendered complete set of source files might be the right thing to search in

mkangia commented on corehq/tabs/tabclasses.py:
```diff
@@ -1340,6 +1340,7 @@ def _get_feature_flag_items(domain):
             'url': reverse(CalendarFixtureConfigView.urlname, args=[domain])
         })
+    # Check change for this
     if toggles.FLAT_LOCATION_FIXTURE.enabled(domain):
```
oops! that was simple.

mkangia committed:
Fix incorrect method call

mkangia committed:
Update condition to reveal Location fixture configuration view

mkangia committed:
fix circular loop for import

czue commented on corehq/apps/locations/management/commands/migrate_new_location_fixture.py:
```diff
@@ -0,0 +1,40 @@
+    domain_with_application = {}
+    for domain in domains:
+        domain_obj = Domain.get_by_name(domain)
+        for application in domain_obj.applications():
```
is there a chance we think we need to search app builds too? maybe just released ones? failure mode I'm thinking of is if someone has removed references in the current app version but still has users using earlier app versions.

mkangia replied:
From the limited knowledge i have, in that case would this stand for all domains that migrate from old fixture format in an earlier version to flat fixture in new versions with users still using old version of app? May be because phone restore would still give new flat fixture format and app would be using old references?

czue replied:
yeah exactly. so for example for TB we have eliminated hierarchical from our apps, but we have not turned off sync yet because one person is still using the old app

czue commented on corehq/apps/locations/management/commands/migrate_new_location_fixture.py:
```diff
@@ -0,0 +1,40 @@
+        for application in domain_obj.applications():
+            raw_doc = json.dumps(application.get_db().get(application.id))
+            if search_string in raw_doc:
+                search_string[domain] = application.id
```
I think this should be `domain_with_application` not `search_string`

czue reviewed (changes requested):
I think we need to be very careful with rolling this out. The failure mode here is that we unknowingly break a huge amount of people who are trying to use their applications. As-is if this PR is merged it will immediately do that.

I think we need to come up with a clear migration pathway that we can feel very confident is not going to break anyone, and I suspect we need to do that in multiple stages. I also think we will want a comprehensive test suite for the logic that checks for location references in applications so we can be sure to trust it.

It seems like it would be much easier to start by just making this the default for new domains, and converting everyone who is not using locations (no org hierarchy defined). Then focusing on figuring out what to do with the hard cases.

mkangia committed:
fix hash formation

esoergel commented:
It's a bit confusing to understand how the `FixtureConfiguration` and the toggles fit together, what is the intended interaction?

I just had an idea - what if you instead worked to drop both toggles and use only the `FixtureConfiguration` to determine what to sync?  This would mean creating a `FixtureConfiguration` for all domains currently using locations (we can be conservative and set `hierarchical=True` at first), then going forward, if a domain doesn't have a `FixtureConfiguration`, we sync only the flat fixture, otherwise we do what the `FixtureConfiguration` says.  That way it's stored clearly in one place, and there's a migration path for existing domains.  (This is just me spitballing, so feel free to ignore if it doesn't sound easier)

czue commented:
@esoergel I like that idea a lot. The only issue is that for any app that is using both we do need to be explicit about which one we resolve in app manager. A couple thoughts:

1. we could add an option to the config and auto-set it safely during the migration
2. we could add as an option inside the app somewhere
3. we could try to detect it on build by inspecting your app contents (might be hard to implement and a performance hit - and we would have to figure out what to do if you referenced none or both of them in your app and likely need to fall back to 1 or 2 to resolve that...)

at a minimum I think making it explicit in the config object as opposed to mucking around with the feature flags will be much easier to reason with.

mkangia committed:
Remove toggle check for flat fixture

This takes care of the case when a domain has HIERARCHICAL_LOCATION_FIXTURE enabled but wants to move to flat fixture, which would have let to no locations since conditions were false for both location structure.

mkangia committed:
Add documentation for methods deciding fixture to pick

mkangia committed:
Re-create migration, Add clear steps

mkangia committed:
Update condition for syncing flat fixture to rely just on flat fixture conf

The default for conf is true for both flat and hierarchical, so must avoid that in case of default conf when its not set

dimagimon commented on corehq/apps/locations/management/commands/migrate_new_location_fixture.py:
```diff
@@ -0,0 +1,59 @@
+        # 4. Update domains that need to stay on hierarchical with enabled legacy toggle to be able to access conf
+        # and update their location configuration to use hierarchical fixture for now
+        toggle = Toggle.get(HIERARCHICAL_LOCATION_FIXTURE.slug)
+        for domain in domains_to_stay_on_hierarchical_fixture :
```
E203 whitespace before ':'

mkangia committed:
fix dimagimon complaint [skip ci]

mkangia commented:
@esoergel @czue have updated spec(Revised Changes V2) according to what was discussed last and also the PR accordingly. I have not tested it as of now. Once confirmed that the approach looks fine, will spend time testing it.
I have not looked in the idea above about dropping both toggles since that needs some more brainstorming.

czue commented:
@mkangia I think it's worth making the chosen format in apps configurable as a part of this work and dropping both toggles. I don't think it should be hard - just use the same logic you are using in the toggles but insert the configuration object in place of it, and set it the same way you are setting the toggles.

czue commented:
so to be clear I think the steps would be:

1. add a field "default format in apps" to `FixtureConfiguration` with allowed values ["flat", "hiearchical"] and default value "flat"
2. replace all toggle logic in apps to instead run off of this field
3. remove any toggle behavior from fixture generation code and rely only on the config object
4. update migration to set this field to "hieararchical" where you are currently setting the toggle.

@esoergel does that sound right?

mkangia commented:
@czue if we drop both toggles, it would make the configuration option open for all users and would allow domains on flat fixture to move back to hierarchical configurations whereas we need the opposite to happen?

czue commented:
good point - we should also hide the config UI if the app type is flat and only flat is turned on (which should also be the default for all new domains)

mkangia commented:
@czue that sounds good. Just so i understand this correctly,
You are suggesting to keep the 2 fields `sync_flat_fixture` and `sync_hierarchical_fixture` as well along with `default format in apps`?

czue commented:
> You are suggesting to keep the 2 fields `sync_flat_fixture` and `sync_hierarchical_fixture` as well along with `default format in apps`? 

yep exactly

esoergel commented:
This all sounds great to me, I'm on the same page.

mkangia committed:
fix and add test case for ensuring correct location instance

mkangia committed:
fix and add test cases for syncing correct location fixture format

mkangia committed:
fix dimagimon complaint

esoergel commented:
There are some build failures, my guess is it's mostly just existing tests that need to be updated to reflect the new fixture.

mkangia committed:
fix test cases relying on hierarchical structure by enabling HIERARCHICAL_LOCATION_FIXTURE feature flag

dimagimon commented on corehq/ex-submodules/casexml/apps/phone/tests/test_ota_fixtures.py:
```diff
@@ -13,6 +13,7 @@
 from corehq.apps.users.dbaccessors.all_commcare_users import delete_all_users
 from casexml.apps.case.tests.util import check_xml_line_by_line
 from corehq.form_processor.tests.utils import run_with_all_backends
+from corehq.util.test_utils import flag_enabled
```
F401 'flag_enabled' imported but unused

mkangia commented:
@esoergel I had a quick sync up with @czue and we concluded to stay on the approach that PR is following. I am having a tough time resolving the build failures. Needs a better knowledge of fixtures to fix them. Moreover, some build failures might highlight genuine concerns. Can you help with test cases? i have fixed  some of them but got stuck at some tricky ones now.

esoergel committed:
Revert "fix test cases..." for non-location tests

esoergel committed:
Only sync flat fixture if the domain uses locations

esoergel committed:
Update test to flat fixture for default

esoergel committed:
Avoid passing unneeded parameter from mock

mkangia commented:
Thank You @esoergel. :1st_place_medal:

esoergel reviewed (commented):
I just did another review and it looks good to me.  Was there anything else you wanted to do before wrapping this up?  (The "don't merge" flag is still on).  @czue did you have any remaining followup?

mkangia commented:
thanks @esoergel One last thing would be to ensure the places where `commtrack:location` exists in the code. I would assume that this notation should be used only for migration and other wise the new notation should be used everywhere.
As of now its mostly in test cases and just one place in code at https://github.com/dimagi/commcare-hq/blob/mk-243262-new-location-fixture/corehq/apps/app_manager/detail_screen.py#L614 which would be good check. Relying on your better understanding of fixtures. Let me know if any code/test-case i should specifically check still using the old notation.

esoergel commented:
That looks like location xpath validation in the app builder, which I don't think I've ever seen in action.  @czue it looks like you wrote that back in 2014, any idea if that still works or where/how I'd be able to try it out?  [This](https://github.com/dimagi/commcare-hq/blob/3ccacef392e3c087ad4eab7c3a85179882d6538e/corehq/apps/app_manager/models.py#L2115) makes me think it's a case list thing, but I'm not sure how to get the option to appear.

czue commented:
@esoergel @mkangia that is what powers [this functionality](https://confluence.dimagi.com/display/commtrack/Referencing+Locations+in+Applications). I think we can just get rid of that functionality if you're on the new fixture format. we probably need to support it for legacy apps though.

czue commented:
I haven't done a thorough review since last time, but I'm on board with the plan and signed off as long as you both are happy. Would be good to address / understand how this affects the behavior of that case list reference code before merging though.

mkangia commented:
Thank You @esoergel @czue.
Will go ahead and try to do the same check for case list reference as well then.

@proteusvacuum Just checking if you did you come across this case list reference when working/moving to flat fixture locations.

czue commented:
> @proteusvacuum Just checking if you did you come across this case list reference when working/moving to flat fixture locations.

there is no way it is currently working for flat fixture so i think we can conclude no one is using it.

mkangia committed:
update location xpath used for flat location fixtures when validating detail column

esoergel commented:
@mkangia I just tried this out and saw that it does indeed still exist and work.  The feature is pretty hidden though.  Go to a module on an app that uses cases (on a domain with locations).  Then add a new property, and type in something like `location:state:district/name` (where `state` and `district` are the codes of two location types on that domain).  When I do that, I get the following error message on save:

> Case List has invalid location reference state:county/name. Details: Reference type county cannot be a child of primary type state.

When I flip the order to `location:district:state/name`, it saves successfully.  If you try it with names that aren't real location types, it will give a different error.

Anyways, I think all we need to do for this migration is turn off that validation if the domain uses the flat fixture.  If I type gibberish in there it pretty much just lets it slide.

czue commented:
thanks @esoergel. sorry i should have mentioned that @mkangia and i offlined about this.

> Anyways, I think all we need to do for this migration is turn off that validation if the domain uses the flat fixture.

I think we should actually just disallow it to use the "location:" syntax at all.

mkangia committed:
Fix check for domains in unexpected state

mkangia committed:
fix issue with setting toggle method call, update and add better tracking to script

mkangia committed:
Revert "update location xpath used for flat location fixtures when validating detail column"

This reverts commit 385a68b501fe73f8bb747cde71121676f7d5a187.

mkangia committed:
remove support for location references in CaseList

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -2096,11 +2096,17 @@ def rename_lang(self, old_lang, new_lang):
+                project = Domain.get_by_name(domain)
                 try:
+                    if not should_sync_hierarchical_fixture(project):
+                        raise LocationXpathValidationError(_('Location references not supported in Case List.'))
```
@czue This would take care of discontinuing the feature to use location references in Case List. Any suggestions for better error message? Any alternative that could be suggested?

czue replied:
Hmm. Maybe "That format is no longer supported. To reference the location hierarchy you need to use the "Custom Calculations in Case List	" feature preview. For more information see: https://confluence.dimagi.com/pages/viewpage.action?pageId=38276915". cc @sheelio

sheelio replied:
I think that message+link is fine.

mkangia committed:
Merge branch 'master' into mk-243262-new-location-fixture

mkangia committed:
Update error message for discontinued feature

esoergel committed:
Drop unneeded arg from test

esoergel commented:
Build failure was unrelated, just kicked it.

mkangia committed:
actually pass the domain when fetching all instances and not just empty string

mkangia committed:
Merge branch 'mk-243262-new-location-fixture' of github.com:dimagi/commcare-hq into mk-243262-new-location-fixture

mkangia committed:
add comment over discontinued feature

mkangia committed:
Merge master

mkangia committed:
update test for add_missing_instance call to pass domain

dimagimon commented on corehq/apps/locations/tests/test_location_fixtures.py:
```diff
@@ -82,6 +85,8 @@ def setUp(self):
         self.user = create_restore_user(self.domain, 'user', '123')
 
+
+    @flag_enabled('HIERARCHICAL_LOCATION_FIXTURE')
```
E303 too many blank lines (2)

mkangia committed:
fix dimagimon complaint [skip ci]

esoergel commented on corehq/apps/locations/management/commands/migrate_new_location_fixture.py:
```diff
@@ -0,0 +1,65 @@
+                # location_conf_for_domain.sync_flat_fixture = True
+                # location_conf_for_domain.sync_hierarchical_fixture = False
+                # location_conf_for_domain.save()
+                print "Domain that needs attention since its not in a definite state: %s" % domain
```
Don't these domains also need the `HIERARCHICAL_LOCATION_FIXTURE` toggle enabled?  I agree that it makes sense to look at them individually, but I imagine in every case we'll want to be able to access the FixtureConfiguration UI and let them keep getting the hierarchical fixture.

mkangia replied:
@esoergel @czue I would expect no domains to be in this area. These would be domains with flat fixture enabled but with configuration set to use hierarchical which contradicts. So this is just a sanity check.
We can always add them to the flag if needed but like i said there should not be any ideally.

mkangia committed:
add check option for migration moving to new location fixture [skip ci]

esoergel commented on corehq/apps/locations/management/commands/migrate_new_location_fixture.py:
```diff
@@ -33,10 +46,26 @@ def handle(self, *args, **options):
+                location_configuration = LocationFixtureConfiguration.for_domain(domain)
+                print "Current Configuration, Persisted: %s, Use Flat fixture: %s, Use Hierarchical fixture %s" % (
+                    not location_configuration._state.adding, location_configuration.sync_hierarchical_fixture,
+                    location_configuration.sync_flat_fixture)
```
Won't the `location_configuration` always be missing for these domains?  They're 
```python
domains_to_stay_on_hierarchical_fixture = (set(domains_having_locations) -
                                           set(domain_with_flat_fixture_enabled))
```
So if the flat fixture toggle isn't enabled, they shouldn't have a fixture config, right?

mkangia replied:
Yup! Thats right. Just logging for confirmation.

mkangia commented:
@esoergel lets do this today!
